java/velox4j/src/main/cpp/main/velox4j/lifecycle/ObjectStore.h

```diff
@@ -0,0 +1,122 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <map>
+#include "ResourceMap.h"
+
+namespace velox4j {
+
+// ObjectHandle is a signed int64 consisting of:
+// 1. 1 - 32 bits is a signed int32 as the object store's ID;
+// 2. 1st bit is always zero to be compatible with jlong;
+// 3. 33 - 64 bits is an unsigned int32 as the object's ID;
+//
+// When the object is tended to be retrieved with its ObjectHandle,
+// the program first finds its resident object store, then looks up
+// for the object in the store.
+using StoreHandle = int32_t;
+using ObjectHandle = int64_t;
+constexpr static ObjectHandle kInvalidObjectHandle = -1;
+
+// A store for caching shared-ptrs and enlarging lifecycles of the ptrs to match
+// lifecycle of the store itself by default, and also serving release calls to
+// release a ptr in advance. This is typically used in JNI scenario to bind a
+// shared-ptr's lifecycle to a Java-side object or some kind of resource
+// manager.
+class ObjectStore {
+ public:
+  static ObjectStore* global() {
+    static std::unique_ptr<ObjectStore> globalStore = create();
+    return globalStore.get();
+  }
+
+  static std::unique_ptr<ObjectStore> create() {
+    static std::mutex mtx;
+    std::lock_guard<std::mutex> lock(mtx);
+    StoreHandle nextId = safeCast<StoreHandle>(stores().nextId());
+    auto store = std::unique_ptr<ObjectStore>(new ObjectStore(nextId));
+    StoreHandle storeId = safeCast<StoreHandle>(stores().insert(store.get()));
+    VELOX_CHECK(
+        storeId == nextId, "Store ID mismatched, this should not happen");
+    return store;
+  }
+
+  static void release(ObjectHandle handle) {
+    ResourceHandle storeId =
+        safeCast<ResourceHandle>(handle >> (sizeof(ResourceHandle) * 8));
+    ResourceHandle resourceId = safeCast<ResourceHandle>(
+        handle & std::numeric_limits<ResourceHandle>::max());
+    auto store = stores().lookup(storeId);
+    store->releaseInternal(resourceId);
+  }
+
+  template <typename T>
+  static std::shared_ptr<T> retrieve(ObjectHandle handle) {
+    ResourceHandle storeId =
+        safeCast<ResourceHandle>(handle >> (sizeof(ResourceHandle) * 8));
+    ResourceHandle resourceId = safeCast<ResourceHandle>(
+        handle & std::numeric_limits<ResourceHandle>::max());
+    auto store = stores().lookup(storeId);
+    return store->retrieveInternal<T>(resourceId);
+  }
+
+  virtual ~ObjectStore();
+
+  StoreHandle id() const {
+    return storeId_;
+  }
+
+  template <typename T>
+  ObjectHandle save(std::shared_ptr<T> obj) {
+    const std::lock_guard<std::mutex> lock(mtx_);
+    const std::string_view description = typeid(T).name();
+    ResourceHandle handle = store_.insert(std::move(obj));
```

Comment:
`store_` is thread-safe. We may move the lock acquisition from line 85 to behind line 87.

Reply:
Good question! It was actually intentional to keep the joint operations on `store_` and `aliveObjects_` atomic. 

See the `release` function:

https://github.com/facebookincubator/velox/pull/13186/files/0d9a73c07803b7aa40aa12754f44676ff3265363#diff-acbc62f2037e212eb63461a33d511eba6d07a304453cd66aca7db054c9cf6579R30-R34

So we have to have `store_` and `aliveObjects_` synchronized at a time, or else inconsistency might be caused when writing and reading happen concurrently.